contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,364 @@
+        // Required.
+        SnapshotHandle string `json:"snapshotHandle"`
+
+        // Timestamp when the point-in-time snapshot is taken on the storage
+        // system. The format of this field should be a Unix nanoseconds time
+        // encoded as an int64. On Unix, the command `date +%s%N` returns
+        // the  current time in nanoseconds since 1970-01-01 00:00:00 UTC.
+        // This field is REQUIRED.
+	CreatedAt int64 `json:"createdAt,omitempty" protobuf:"varint,3,opt,name=createdAt"`
+}
+
+```
+
+#### The `VolumeSnapshotClass` Object
+
+A new VolumeSnapshotClass API object will be added instead of resuing the existing StorageClass, in order to avoid mixing parameters between snapshots and volumes. Each CSI Volume Driver can have its own default VolumeSnapshotClass. If VolumeSnapshotClass is not provided, a default will be used. It allows to add new parameters for snapshots.
+
+```
+
+type VolumeSnapshotClass struct {
+        metav1.TypeMeta `json:",inline"`
+        // +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Snapshotter is the driver expected to handle this VolumeSnapshotClass.
+	// This value may not be empty.
+	Snapshotter string
+
+	// Parameters holds parameters for the snapshotter.
+        // These values are opaque to the system and are passed directly
+        // to the snapshotter. The only validation done on keys is that they are
+        // not empty. The maximum number of parameters is
+        // 512, with a cumulative max size of 256K
+        // +optional
+	Parameters map[string]string
+}
+
+
+```
+### Volume API Changes
+
+With Snapshot API avaialble, users could provision volumes from snapshot and data will be prepopulated to the volumes. Also considering clone and other possible storage operations, there could be many different types of sources used for populating the data to the volumes. In this proposal, we add a general "DataSource" which could be used to represent different types of data sources.
+
+#### The `DataSource` Object in PVC
+
+Add a new `DataSource` field into both PVC and PV to represent the source of the data which is prepopulated to the provisioned volume. If an external-provisioner does not understand the new `DataSource` field and does not pre-populate the new volume, PV/PVC controller should be able to detect that (e.g. PVC has `DataSource` but PV does not) and fail the operation.
+
+Possible `DataSource` types may include the following:
+
+    * VolumeSnapshot: restore snapshot to a new volume
+    * PersistentVolumeClaim: clone volume which is represented by PVC
+
+```
+
+type PersistentVolumeClaimSpec struct {
+        // If specified, volume will be prepopulated with data from the PVCDataSourceRef.
+        // +optional
+        DataSourceRef *TypedLocalObjectReference `json:"dataSourceRef" protobuf:"bytes,2,opt,name=dataSourceRef"`
+}
+
+type PersistentVolumeSpec struct {
```

Comment:
Do we really need this?  Is it not sufficient for it to be in the PVC?

if so, same comments as above

Reply:
What if the PVC and PV have different sources?

Reply:
Is this going to be used as part of the PV<->PVC match logic?